src/SourceBuild/content/eng/tools/BinaryToolKit/BinaryTool.CompareBinariesAgainstBaselines.cs

```diff
@@ -0,0 +1,130 @@
+{
+    private const string PatternsToRemoveMarker = "# Consider removing the following unused patterns from the baseline file:";
+
+    private List<string> CompareBinariesAgainstBaselines(
+        IEnumerable<string> detectedBinaries,
+        string? allowedBinariesFile,
+        string? disallowedSbBinariesFile,
+        string outputReportDirectory,
+        string targetDirectory,
+        Modes mode)
+    {
+        Log.LogInformation("Comparing detected binaries to baseline(s).");
+
+        var binariesToRemove = GetUnmatchedBinaries(
+            detectedBinaries,
+            allowedBinariesFile,
+            outputReportDirectory,
+            targetDirectory,
+            mode).ToList();
+
+        if (mode.HasFlag(Modes.Validate))
+        {
+            var nonSbBinariesToRemove = GetUnmatchedBinaries(
+                detectedBinaries,
+                disallowedSbBinariesFile,
+                outputReportDirectory,
+                targetDirectory,
+                mode).ToList();
+    
+            var newBinaries = binariesToRemove.Intersect(nonSbBinariesToRemove);
+
+            if (newBinaries.Any())
+            {
+                string newBinariesFile = Path.Combine(outputReportDirectory, "NewBinaries.txt");
+
+                File.WriteAllLines(newBinariesFile, newBinaries);
+
+                Log.LogWarning($"    {newBinaries.Count()} new binaries detected. Check {newBinariesFile}");
+            }
+        }
+
+        Log.LogInformation("Finished comparing binaries.");
+
+        return binariesToRemove;
+    }
+
+    private IEnumerable<string> GetUnmatchedBinaries(
+        IEnumerable<string> searchFiles,
+        string? baselineFile,
+        string outputReportDirectory,
+        string targetDirectory,
+        Modes mode)
+    {
+        var patterns = ParseBaselineFile(baselineFile);
+
+        if (mode.HasFlag(Modes.Validate))
+        {
+            // If validating in any mode (Mode == Validate or Mode == All), 
+            // we need to detect both unused patterns and unmatched files.
+            // We simultaneously detect unused patterns and unmatched files for efficiency.
+
+            HashSet<string> unusedPatterns = new HashSet<string>(patterns);
+            HashSet<string> unmatchedFiles = new HashSet<string>(searchFiles);
+
+            foreach (string pattern in patterns)
+            {
+                Matcher matcher = new Matcher(StringComparison.Ordinal);
+                matcher.AddInclude(pattern);
+                
+                var matches = matcher.Match(targetDirectory, searchFiles);
+                if (matches.HasMatches)
+                {
+                    unusedPatterns.Remove(pattern);
+                    unmatchedFiles.ExceptWith(matches.Files.Select(file => file.Path));
+                }
+            }
+
+            UpdateBaselineFile(baselineFile, outputReportDirectory, unusedPatterns);
+
+            return unmatchedFiles;
+        }
+        else
```

Comment:
A good practice IMO is to check the mode state here and not assume the mode.  This if/when a new option is every added.